File: Replace_word.py
// Time Complexity : O(m*n + m1 * n1)
            'm' words, 'n' avg len of the word, 
            'm1' words in sentence, 'n1' avg len of word in setence,
 // Space Complexity : O(w)
            'w' is the no of unique characters in all words in dic
class TrieNode:
    def __init__(self):
        self.children=[None]*26
        self.word=None
class Solution:
    def __init__(self):
        self.root=TrieNode()
    def insert(self,word):
        curr=self.root
        for i in word:
            ch=ord(i)-ord('a')
            if not curr.children[ch]:
                curr.children[ch]=TrieNode()
            curr=curr.children[ch]
        curr.word=word
        
    def search(self,word):
        curr=self.root
        for i in word:
            ch=ord(i)-ord('a')
            if curr.children[ch]:
                curr=curr.children[ch]
                if curr.word:
                    return curr.word
            else:
                break
        return word
    
    def replaceWords(self, dict: List[str], sentence: str) -> str:
        for d in dict:
            self.insert(d)
            
        output=''
        for word in sentence.split():
            if len(output) > 0:
                output+=' '
            output+= self.search(word)
        return output

File: ReplaceWord.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Trie
{
    public class ReplaceWord
    {
        /*
         * T.C & S.C O(M*L) + O(N*K) 
         * where M is number of words in the sentence and L is avg length of each word 
         * N is number of words in disctionary and N is avg length of each word 
         * 
         */
        public class TrieNode
        {
            public TrieNode[] children;
            public bool isEnd;

            public TrieNode()
            {
                children = new TrieNode[26];
            }
        }

        TrieNode root = new TrieNode();

        private void Insert(string word)
        {
            TrieNode curr = root;

            for (int i = 0; i < word.Length; i++)
            {
                char c = word[i];

                if (curr.children[c - 'a'] == null)
                {
                    curr.children[c - 'a'] = new TrieNode();
                }
                curr = curr.children[c - 'a'];
            }
            curr.isEnd = true;

        }

        public string ReplaceWords(IList<string> dictionary, string sentence)
        {

            StringBuilder result = new StringBuilder();

            foreach (string word in dictionary)
            {
                Insert(word);
            }

            string[] strArray = sentence.Split(" ");

            for (int i = 0; i < strArray.Length; i++)
            {
                if (i != 0)
                {
                    result.Append(" ");
                }
                TrieNode curr = root;

                string word = strArray[i];

                StringBuilder newStr = new StringBuilder();

                for (int j = 0; j < word.Length; j++)
                {
                    char c = word[j];
                    if (curr.children[c - 'a'] == null || curr.isEnd)
                    {
                        break;
                    }
                    curr = curr.children[c - 'a'];
                    newStr.Append(c);
                }

                if (curr.isEnd)
                {
                    result.Append(newStr.ToString());
                }
                else
                {
                    result.Append(word);
                }

            }
            return result.ToString();

        }
    }
}
